features/filesystem/sd/SDBlockDevice.cpp

```diff
@@ -0,0 +1,581 @@
+            return BD_ERROR_OK;
+        }
+    }
+
+    debug_if(_dbg, "Timeout waiting for v1.x card\n");
+    return BD_ERROR_DEVICE_ERROR;
+}
+
+bd_error_t SDBlockDevice::initialise_card_v2()
+{
+    for (int i = 0; i < SD_COMMAND_TIMEOUT; i++) {
+        wait_ms(50);
+        _cmd58();
+        _cmd(55, 0);
+        if (_cmd(41, 0x40000000) == 0) {
+            _cmd58();
+            debug_if(_dbg, "\n\rInit: SDCARD_V2\n\r");
+            cdv = 1;
+            return BD_ERROR_OK;
+        }
+    }
+
+    debug_if(_dbg, "Timeout waiting for v2.x card\n");
+    return BD_ERROR_DEVICE_ERROR;
+}
+
+bd_error_t SDBlockDevice::init()
+{
+    _lock.lock();
+    bd_error_t err = initialise_card();
+    _is_initialized = (err == BD_ERROR_OK);
+    if (!_is_initialized) {
+        debug_if(_dbg, "Fail to initialize card\n");
+        _lock.unlock();
+        return err;
+    }
+    debug_if(_dbg, "init card = %d\n", _is_initialized);
+    _sectors = _sd_sectors();
+
+    // Set block length to 512 (CMD16)
+    if (_cmd(16, 512) != 0) {
+        debug_if(_dbg, "Set 512-byte block timed out\n");
+        _lock.unlock();
+        return BD_ERROR_DEVICE_ERROR;
+    }
+
+    // Set SCK for data transfer
+    _spi.frequency(_transfer_sck);
+    _lock.unlock();
+    return BD_ERROR_OK;
+}
+
+bd_error_t SDBlockDevice::deinit()
+{
+}
+
+bd_error_t SDBlockDevice::program(const void *b, bd_addr_t addr, bd_size_t size)
+{
+    if (!is_valid_program(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    _lock.lock();
+    if (!_is_initialized) {
+        _lock.unlock();
+        return BD_ERROR_NO_INIT;
+    }
+
+    const uint8_t *buffer = static_cast<const uint8_t*>(b);
+    while (size > 0) {
+        bd_addr_t block = addr / 512;
+        // set write address for single block (CMD24)
+        if (_cmd(24, block * cdv) != 0) {
+            _lock.unlock();
+            return BD_ERROR_DEVICE_ERROR;
+        }
+
+        // send the data block
+        _write(buffer, 512);
+        buffer += 512;
+        addr += 512;
+        size -= 512;
+    }
+    _lock.unlock();
+    return 0;
+}
+
+bd_error_t SDBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    if (!is_valid_read(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    _lock.lock();
+    if (!_is_initialized) {
+        _lock.unlock();
+        return BD_ERROR_PARAMETER;
+    }
+    
+    uint8_t *buffer = static_cast<uint8_t *>(b);
+    while (size > 0) {
+        bd_addr_t block = addr / 512;
+        // set read address for single block (CMD17)
+        if (_cmd(17, block * cdv) != 0) {
+            _lock.unlock();
+            return BD_ERROR_DEVICE_ERROR;
+        }
+        
+        // receive the data
+        _read(buffer, 512);
+        buffer += 512;
+        addr += 512;
+        size -= 512;
+    }
+    _lock.unlock();
+    return 0;
+}
+
+bd_error_t SDBlockDevice::erase(bd_addr_t addr, bd_size_t size)
+{
+    return 0;
```

Comment:
As for why the SD class can't just implement write: There's no good way in c++ to have mutually recursive abstract functions, that is, you can't give an implementor the option of implementing `write` or `program/erase`. Since we can't give the implementor a choice, I went with having the more power functions abstract, since we want to encourage implementors to expose the most functionality.